src/ray/core_worker/test/actor_manager_test.cc

```diff
@@ -0,0 +1,270 @@
+  ActorID actor_id = ActorID::Of(job_id, task_id, 1);
+  ActorHandle *actor_handle_that_doesnt_exist = nullptr;
+  ASSERT_TRUE(actor_manager_->GetActorHandle(actor_id, &actor_handle_that_doesnt_exist)
+                  .IsInvalid());
+}
+
+TEST_F(ActorManagerTest, TestSerializeAndDeserializeActorHandles) {
+  JobID job_id = JobID::FromInt(1);
+  const TaskID task_id = TaskID::ForDriverTask(job_id);
+  ActorID actor_id = ActorID::Of(job_id, task_id, 1);
+  const auto caller_address = rpc::Address();
+  const auto call_site = "";
+  EXPECT_CALL(*actor_info_accessor_, AsyncSubscribe(_, _, _))
+      .WillRepeatedly(testing::Return(Status::OK()));
+  EXPECT_CALL(*reference_counter_, SetDeleteCallback(_, _))
+      .WillRepeatedly(testing::Return(true));
+
+  // Make sure serialization fails when actor handle is not registered
+  std::string serialized_bytes;
+  ObjectID actor_handle_id;
+  ASSERT_TRUE(
+      actor_manager_->SerializeActorHandle(actor_id, &serialized_bytes, &actor_handle_id)
+          .IsInvalid());
+  // Add an actor handle.
+  std::shared_ptr<ActorHandle> actor_handle = CreateActorHandle(actor_id, job_id);
+  ASSERT_TRUE(actor_manager_->AddActorHandle(move(actor_handle), true, task_id, call_site,
+                                             caller_address));
+  // Serialize actor_handle first
+  ASSERT_TRUE(
+      actor_manager_->SerializeActorHandle(actor_id, &serialized_bytes, &actor_handle_id)
+          .ok());
+  // Make sure deserialization & register works.
+  ObjectID outer_object_id = ObjectID::Nil();
+  // Sinece DeserializeAndRegisterActorHandle happens in a non-owner worker, we should
+  // make sure it borrows an object.
+  EXPECT_CALL(*reference_counter_, AddBorrowedObject(_, _, _, _));
+  ActorID returned_actor_id = actor_manager_->DeserializeAndRegisterActorHandle(
+      serialized_bytes, outer_object_id, task_id, call_site, caller_address);
+  ASSERT_TRUE(returned_actor_id == actor_id);
+  // Let's try to get the handle and make sure it works.
+  ActorHandle *actor_handle_to_get = nullptr;
+  ASSERT_TRUE(actor_manager_->GetActorHandle(actor_id, &actor_handle_to_get).ok());
+  ASSERT_TRUE(actor_handle_to_get->GetActorID() == actor_id);
+  ASSERT_TRUE(actor_handle_to_get->CreationJobID() == job_id);
+}
+
+TEST_F(ActorManagerTest, TestActorStateNotificationPending) {
+  JobID job_id = JobID::FromInt(1);
+  const TaskID task_id = TaskID::ForDriverTask(job_id);
+  ActorID actor_id = ActorID::Of(job_id, task_id, 1);
+
+  // Nothing happens if state is pending.
+  EXPECT_CALL(*direct_actor_submitter_, ConnectActor(_, _)).Times(0);
+  EXPECT_CALL(*direct_actor_submitter_, DisconnectActor(_, _)).Times(0);
+  rpc::ActorTableData actor_table_data;
+  actor_table_data.set_actor_id(actor_id.Binary());
+  actor_table_data.set_state(
+      rpc::ActorTableData_ActorState::ActorTableData_ActorState_PENDING);
+  actor_manager_->HandleActorStateNotification(actor_id, actor_table_data);
```

Comment:
These tests look good, but it would be even nicer if we could test the full flow from adding an actor handle to receiving the async notification, instead calling `HandleActorStateNotification` directly. One way to do that is to queue the `HandleActorStateNotification` callback that gets scheduled when the actor handle is added, then choose when to call the callback in the unit test.

Reply:
That sounds nice actually. It sounds like it will help us testing complex race conditions for cluster hangs PR as well. I will try!